pep-0483.txt

```diff
@@ -38,276 +70,285 @@ Summary of gradual typing
+        return first if len(first) >= len(second) else second
 
-  * def longest(a: AnyStr, b: AnyStr) -> AnyStr:
+    result = longest('a', 'abc')
 
-    -  return a if len(a) >= len(b) else b
+  The inferred type of ``result`` is still ``Union[str, bytes]``, even though
+  both arguments are ``str``.
 
-  * x = longest('a', 'abc')  # The inferred type for x is str
 
-  * y = longest('a', b'abc')  # Fails static type check
+Defining and using generic types
+--------------------------------
 
-  * In this example, both arguments to longest() must have the same type
-    (str or bytes), and moreover, even if the arguments are instances of a
-    common str subclass, the return type is still str, not that subclass
-    (see next example).
+- ``class MyGeneric(Generic[X, Y, ...]):`` ... Define a generic type
+  ``MyGeneric`` over type variables ``X``, etc. ``MyGeneric`` itself becomes
+  parameterizable, e.g. ``MyGeneric[int, str, ...]`` is a specific type with
+  substitutions ``X->int``, etc.
+
+- TODO: Explain use of generic types in function signatures. E.g.
+  ``Sequence[X]``, ``Sequence[int]``, ``Sequence[Tuple[X, Y, Z]]``,
+  and mixtures.
+  Think about co\*variance. No gimmicks like deriving from
+  ``Sequence[Union[int, str]]`` or ``Sequence[Union[int, X]]``.
+  Add notes on indexed generics
 
-- For comparison, if the type variable was unconstrained, the common
-  subclass would be chosen as the return type, e.g.:
 
-  * S = TypeVar('S')
+Covariance and Contravariance
+-----------------------------
 
-  * def longest(a: S, b: S) -> S:
+TODO: Definition, declaring variance, example with functions,
+covatiant example with ``Tuple``, contravariant example with ``Callable``,
+mutable types, why ``List`` is invariant.
 
-    -  return a if len(a) >= b else b
+One of the best examples to illustrate contravariance is callable type.
+It is covariant in the return type, but contravariant in the
+arguments. "Covariant" here means that for two callable types that
+differ only in the return type, the subtype relationship for the
+callable types follows that of the return types. (Example:
+``Callable[[], Manager]`` is a subtype of ``Callable[[], Employee]``.)
+"Contravariant" here means that for two callable types that differ
+only in the type of one argument, the subtype relationship for the
+callable types goes in the opposite direction as for the argument
+types. (Example: ``Callable[[Employee], None]`` is a subtype of
+``Callable[[Manager], None]``. Yes, you read that right.)
 
-  * class MyStr(str): ...
 
-  * x = longest(MyStr('a'), MyStr('abc'))
 
-  * The inferred type of x is MyStr (whereas in the AnyStr example it would
-    be str).
+Pragmatics
+==========
 
-- Also for comparison, if a Union is used, the return type also has to be
-  a Union:
+Some things are irrelevant to the theory but make practical use more
+convenient. (This is not a full list; I probably missed a few and some
+are still controversial or not fully specified.)
```

Comment:
In a sense, the way you have to declare type variables using `TypeVar(...)` is also part of this.